gnupg-pkcs11-scd/keyutil.c

```diff
@@ -313,3 +435,118 @@ char *keyutil_get_cert_hexgrip (gcry_sexp_t sexp)
 
 	return ret;
 }
+
+void keyinfo_data_free(keyinfo_data_list list) {
+	keyinfo_data_list next, curr;
+
+	if (list == NULL) {
+		return;
+	}
+
+	for (curr = list; curr != NULL; curr = next) {
+		next = curr->next;
+
+		if (curr->value != NULL) {
+			if (curr->value_free != NULL) {
+				curr->value_free(curr->value);
+			}
+		}
+
+		if (curr->tag != NULL) {
+			if (curr->tag_free != NULL) {
+				curr->tag_free(curr->tag);
+			}
+		}
+
+		free(curr);
+	}
+}
+
+keyinfo_data_list keyinfo_get_key_data(keyinfo keyinfo) {
+	keyinfo_data_list first = NULL, n_item = NULL, e_item = NULL;
+	unsigned char *n_hex = NULL;
+	unsigned char *e_hex = NULL;
+
+	if (keyinfo->type == KEYINFO_KEY_TYPE_INVALID) {
+		return NULL;
+	}
+
+	if (keyinfo->type != KEYINFO_KEY_TYPE_UNKNOWN) {
```

Comment:
I truly think you should drop one of the invalid and unknown, I see no benefit of having them both as there is no different logic.

Reply:
The `INVALID` case used to `abort()` because a use-after-free had been detected, but now it must always do the same thing as `UNKNOWN`

Reply:
no `abort()` allowed in code, a proper error should be raised and resource management should take that into account.